### PR TITLE
[QoS] Fix/qos system table rate limiting

### LINKED ISSUE #2734: [QoS] Transactions table rate limiting
Transaction table reads should not get rate limited. The commit step can fail even though we have successfully read/written to the data tables.

One of the causes here is that the number of estimated bytes can be huge for the `_transaction` table and the adjustment will be a negative but the transaction fails because the `_transaction` table read is rate limited.

```
get_range_slices from the table default.todo with slice SlicePredicate(slice_range:SliceRange(start:, finish:, reversed:false, count:2147483647)) and keys KeyRange(start_key:, end_key:, count:110)
Consuming estimated bytes of 11000
Recording adjusted bytes of 104500
multiget_slice from the table _transactions with slice SlicePredicate(slice_range:SliceRange(start:00 01 74 00 00 08 FF FF FF FF FF FF FF FF 00, finish:00 01 74 00 00 08 00 00 00 00 00 00 00 00 00, reversed:false, count:1)) and keys 108
Consuming estimated bytes of 10800
12:49:59.055 [Atlas loadWithTs 108 cells from _transactions on /127.0.0.1:33080 main calling cassandra host /127.0.0.1:33080 started at 2017-11-22T12:49:59.054Z - 970] WARN  c.p.a.k.c.CassandraClientPoolingContainer - Error occurred talking to host '127.0.0.1': com.palantir.atlasdb.qos.ratelimit.RateLimitExceededException: Rate limited. Available capacity has been exhausted.
12:49:59.062 [main] WARN  c.p.a.t.i.AbstractTransactionManager - [605f8e5f-0985-4169-afd8-0e53aba85a1a] RuntimeException while processing transaction.
com.palantir.atlasdb.qos.ratelimit.RateLimitExceededException: Rate limited. Available capacity has been exhausted.
	at com.palantir.atlasdb.qos.ratelimit.QosRateLimiter.consumeWithBackoff(QosRateLimiter.java:84) ~[qos-service-impl/:na]
	at com.palantir.atlasdb.qos.client.AtlasDbQosClient.execute(AtlasDbQosClient.java:77) ~[qos-service-impl/:na]
	at com.palantir.atlasdb.qos.client.AtlasDbQosClient.executeRead(AtlasDbQosClient.java:57) ~[qos-service-impl/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.qos.QosCassandraClient.multiget_slice(QosCassandraClient.java:72) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.lambda$multigetInternal$13(CassandraKeyValueServiceImpl.java:1241) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.TracingQueryRunner.run(TracingQueryRunner.java:50) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.TracingQueryRunner.run(TracingQueryRunner.java:59) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.multigetInternal(CassandraKeyValueServiceImpl.java:1241) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.access$300(CassandraKeyValueServiceImpl.java:156) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl$2.apply(CassandraKeyValueServiceImpl.java:676) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl$2.apply(CassandraKeyValueServiceImpl.java:654) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraClientPoolingContainer.runWithGoodResource(CassandraClientPoolingContainer.java:128) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraClientPoolingContainer.runWithPooledResource(CassandraClientPoolingContainer.java:104) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraClientPoolImpl.runWithPooledResourceRecordingMetrics(CassandraClientPoolImpl.java:646) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraClientPoolImpl.runWithRetryOnHost(CassandraClientPoolImpl.java:598) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.AutoDelegate_CassandraClientPool.runWithRetryOnHost(AutoDelegate_CassandraClientPool.java:18) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.lambda$getLoadWithTsTasksForSingleHost$6(CassandraKeyValueServiceImpl.java:653) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.util.AnnotatedCallable.call(AnnotatedCallable.java:54) ~[atlasdb-client/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.runAllTasksCancelOnFailure(CassandraKeyValueServiceImpl.java:2474) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.loadWithTs(CassandraKeyValueServiceImpl.java:621) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.get(CassandraKeyValueServiceImpl.java:578) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.get(CassandraKeyValueServiceImpl.java:558) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.cassandra.AutoDelegate_CassandraKeyValueService.get(AutoDelegate_CassandraKeyValueService.java:160) ~[atlasdb-cassandra/:na]
	at com.palantir.atlasdb.keyvalue.impl.ProfilingKeyValueService.lambda$get$14(ProfilingKeyValueService.java:180) ~[atlasdb-client/:na]
	at com.palantir.atlasdb.logging.KvsProfilingLogger.maybeLog(KvsProfilingLogger.java:119) ~[atlasdb-client/:na]
	at com.palantir.atlasdb.keyvalue.impl.ProfilingKeyValueService.get(ProfilingKeyValueService.java:180) ~[atlasdb-client/:na]
	at com.palantir.atlasdb.keyvalue.impl.ForwardingKeyValueService.get(ForwardingKeyValueService.java:95) ~[atlasdb-client/:na]
	at com.palantir.atlasdb.keyvalue.impl.TracingKeyValueService.get(TracingKeyValueService.java:183) ~[atlasdb-client/:na]
	at sun.reflect.GeneratedMethodAccessor86.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_92]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_92]
	at com.palantir.tritium.proxy.InvocationEventProxy.execute(InvocationEventProxy.java:140) ~[tritium-lib-0.8.4.jar:0.8.4]
	at com.palantir.tritium.proxy.InvocationEventProxy.instrumentInvocation(InvocationEventProxy.java:117) ~[tritium-lib-0.8.4.jar:0.8.4]
	at com.palantir.tritium.proxy.InvocationEventProxy.handleInvocation(InvocationEventProxy.java:97) ~[tritium-lib-0.8.4.jar:0.8.4]
	at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:87) ~[guava-18.0.jar:na]
	at com.sun.proxy.$Proxy13.get(Unknown Source) ~[na:na]
	at com.palantir.atlasdb.keyvalue.impl.ValidatingQueryRewritingKeyValueService.get(ValidatingQueryRewritingKeyValueService.java:138) ~[atlasdb-client/:na]
	at com.palantir.atlasdb.transaction.service.SimpleTransactionService.get(SimpleTransactionService.java:61) ~[atlasdb-impl-shared/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_92]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_92]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_92]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_92]
	at com.palantir.tritium.proxy.InvocationEventProxy.execute(InvocationEventProxy.java:140) ~[tritium-lib-0.8.4.jar:0.8.4]
	at com.palantir.tritium.proxy.InvocationEventProxy.instrumentInvocation(InvocationEventProxy.java:117) ~[tritium-lib-0.8.4.jar:0.8.4]
	at com.palantir.tritium.proxy.InvocationEventProxy.handleInvocation(InvocationEventProxy.java:97) ~[tritium-lib-0.8.4.jar:0.8.4]
	at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:87) ~[guava-18.0.jar:na]
	at com.sun.proxy.$Proxy15.get(Unknown Source) ~[na:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransaction.loadCommitTimestamps(SnapshotTransaction.java:1830) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransaction.getCommitTimestamps(SnapshotTransaction.java:1810) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransaction.getWithPostFilteringInternal(SnapshotTransaction.java:1056) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransaction.getWithPostFiltering(SnapshotTransaction.java:1042) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransaction.postFilterRows(SnapshotTransaction.java:999) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransaction.access$500(SnapshotTransaction.java:158) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransaction$6.computeNext(SnapshotTransaction.java:902) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransaction$6.computeNext(SnapshotTransaction.java:894) ~[atlasdb-impl-shared/:na]
	at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:143) ~[guava-18.0.jar:na]
	at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:138) ~[guava-18.0.jar:na]
	at com.google.common.collect.Iterators$5.hasNext(Iterators.java:547) ~[guava-18.0.jar:na]
	at com.palantir.common.base.ClosableIterators$1.hasNext(ClosableIterators.java:32) ~[atlasdb-commons/:na]
	at com.google.common.collect.ForwardingIterator.hasNext(ForwardingIterator.java:43) ~[guava-18.0.jar:na]
	at com.google.common.collect.Iterators$PeekingImpl.hasNext(Iterators.java:1149) ~[guava-18.0.jar:na]
	at com.palantir.common.collect.IteratorUtils$3.computeNext(IteratorUtils.java:154) ~[atlasdb-commons/:na]
	at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:143) ~[guava-18.0.jar:na]
	at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:138) ~[guava-18.0.jar:na]
	at com.google.common.collect.TransformedIterator.hasNext(TransformedIterator.java:43) ~[guava-18.0.jar:na]
	at com.google.common.collect.Iterators$7.computeNext(Iterators.java:650) ~[guava-18.0.jar:na]
	at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:143) ~[guava-18.0.jar:na]
	at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:138) ~[guava-18.0.jar:na]
	at com.google.common.collect.ForwardingIterator.hasNext(ForwardingIterator.java:43) ~[guava-18.0.jar:na]
	at com.google.common.collect.ForwardingIterator.hasNext(ForwardingIterator.java:43) ~[guava-18.0.jar:na]
	at com.google.common.collect.Iterators$6.hasNext(Iterators.java:617) ~[guava-18.0.jar:na]
	at com.palantir.common.base.BatchingVisitableFromIterable.batchAcceptSizeHint(BatchingVisitableFromIterable.java:59) ~[atlasdb-commons/:na]
	at com.palantir.common.base.AbstractBatchingVisitable.batchAccept(AbstractBatchingVisitable.java:40) ~[atlasdb-commons/:na]
	at sun.reflect.GeneratedMethodAccessor35.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_92]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_92]
	at com.palantir.common.proxy.SingleCallProxy.invoke(SingleCallProxy.java:48) ~[atlasdb-commons/:na]
	at com.sun.proxy.$Proxy27.batchAccept(Unknown Source) ~[na:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransaction.getBatchingVisitableFromIterator(SnapshotTransaction.java:839) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransaction.access$400(SnapshotTransaction.java:158) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransaction$5.batchAcceptSizeHint(SnapshotTransaction.java:815) ~[atlasdb-impl-shared/:na]
	at com.palantir.common.base.AbstractBatchingVisitable.batchAccept(AbstractBatchingVisitable.java:44) ~[atlasdb-commons/:na]
	at com.palantir.atlasdb.transaction.impl.SerializableTransaction$2.batchAccept(SerializableTransaction.java:214) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.ete.QosCassandraReadTestSuite.lambda$readOneBatchOfSize$7(QosCassandraReadTestSuite.java:189) [atlasdb-ete-tests/:na]
	at com.palantir.atlasdb.transaction.api.LockAwareTransactionTasks$1.execute(LockAwareTransactionTasks.java:32) ~[atlasdb-lock-api/:na]
	at com.palantir.atlasdb.transaction.api.LockAwareTransactionTasks.lambda$asLockUnaware$0(LockAwareTransactionTasks.java:49) ~[atlasdb-lock-api/:na]
	at com.palantir.atlasdb.transaction.impl.AbstractTransactionManager.runTaskThrowOnConflict(AbstractTransactionManager.java:107) ~[atlasdb-client/:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransactionManager.finishRunTaskWithLockThrowOnConflict(SnapshotTransactionManager.java:161) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.SerializableTransactionManager.finishRunTaskWithLockThrowOnConflict(SerializableTransactionManager.java:38) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.SnapshotTransactionManager.runTaskWithLocksThrowOnConflict(SnapshotTransactionManager.java:133) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.SerializableTransactionManager.runTaskWithLocksThrowOnConflict(SerializableTransactionManager.java:38) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.AbstractLockAwareTransactionManager.runTaskThrowOnConflict(AbstractLockAwareTransactionManager.java:128) ~[atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.transaction.impl.AbstractTransactionManager.runTaskWithRetry(AbstractTransactionManager.java:60) ~[atlasdb-client/:na]
	at com.palantir.atlasdb.transaction.impl.AutoDelegate_SerializableTransactionManager.runTaskWithRetry(AutoDelegate_SerializableTransactionManager.java:142) [atlasdb-impl-shared/:na]
	at com.palantir.atlasdb.ete.QosCassandraReadTestSuite.readOneBatchOfSize(QosCassandraReadTestSuite.java:184) [atlasdb-ete-tests/:na]
	at com.palantir.atlasdb.ete.QosCassandraReadTestSuite.shouldBeAbleToReadLargeAmountsExceedingTheLimitFirstTime(QosCassandraReadTestSuite.java:173) [atlasdb-ete-tests/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_92]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_92]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_92]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_92]
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) [junit-4.12.jar:4.12]
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325) [junit-4.12.jar:4.12]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78) [junit-4.12.jar:4.12]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) [junit-4.12.jar:4.12]
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) [junit-4.12.jar:4.12]
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48) [junit-4.12.jar:4.12]
	at org.junit.rules.RunRules.evaluate(RunRules.java:20) [junit-4.12.jar:4.12]
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363) [junit-4.12.jar:4.12]
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137) [junit-4.12.jar:4.12]
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68) [junit-rt.jar:na]
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51) [junit-rt.jar:na]
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242) [junit-rt.jar:na]
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70) [junit-rt.jar:na]
	Suppressed: com.palantir.atlasdb.util.SuppressedException: RuntimeException [com.palantir.atlasdb.qos.ratelimit.RateLimitExceededException: Rate limited. Available capacity has been exhausted.] occurred while processing thread (Atlas loadWithTs 108 cells from _transactions on /127.0.0.1:33080 main)
		... 122 common frames omitted

com.palantir.atlasdb.qos.ratelimit.RateLimitExceededException: Rate limited. Available capacity has been exhausted.

	at com.palantir.atlasdb.qos.ratelimit.QosRateLimiter.consumeWithBackoff(QosRateLimiter.java:84)
	at com.palantir.atlasdb.qos.client.AtlasDbQosClient.execute(AtlasDbQosClient.java:77)
```

### PR DESCRIPTION
**Pull request by hsaraogi**
**Goals (and why)**: Fix #2734.

**Implementation Description (bullets)**: Estimate zero for _transactions/_metadata table for all queries on these tables.

**Where should we start reviewing?**: ThriftQueryWeighers.java

**Priority (whenever / two weeks / yesterday)**: today

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2739)

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/QosCassandraClient.java [hunk 1]
```diff
@@ -40,15 +41,19 @@
...
 import com.palantir.atlasdb.keyvalue.cassandra.CqlQuery;
 import com.palantir.atlasdb.qos.QosClient;
+import com.palantir.atlasdb.transaction.impl.TransactionConstants;
 
 @SuppressWarnings({"all"}) // thrift variable names.
 public class QosCassandraClient implements CassandraClient {
 
     private static final Logger log = LoggerFactory.getLogger(CassandraClient.class);
+    private static final Function<TableReference, Boolean> ZERO_ESTIMATE_DETERMINING_FUNCTION = tRef ->
+            tRef.equals(TransactionConstants.TRANSACTION_TABLE) || tRef.equals(AtlasDbConstants.DEFAULT_METADATA_TABLE);
```
Comment (jeremyk-91) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/QosCassandraClient.java]: Should we exempt persisted locks too? We could try `HiddenTables::isHidden` for the function if you think we should.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java [hunk 2]
```diff
@@ -50,39 +50,50 @@
...
             public QueryWeight estimate() {
-                return ImmutableQueryWeight.builder()
-                        .from(DEFAULT_ESTIMATED_WEIGHT)
-                        .numBytes(ESTIMATED_NUM_BYTES_PER_ROW * numberOfQueriedRows)
-                        .build();
+                if (zeroEstimate) {
+                    return ZERO_ESTIMATED_WEIGHT;
+                }
+                return defaultEstimate();
             }
```
Comment (jeremyk-91) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java]: I'd suggest having two weighers, one that always estimates at 0 and one that returns the default estimate (the idea being that to keep the weighers simple) and having this method pick which one to create. (Not blocking at all though!)

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighersTest.java [hunk 3]
```diff
@@ -77,26 +97,54 @@ public void multigetSliceWeigherReturnsCorrectNumRows() {
...
+        assertThat(numBytesWithTwoRows).isEqualTo(zeroEstimate ? 0L : numBytesWithOneRow * 2);
     }
 
     @Test
     public void rangeSlicesWeigherReturnsCorrectNumRows() {
+        assertThatrangeSlicesWeigherReturnsCorrectNumRows(ThriftQueryWeighers.getRangeSlices(new KeyRange(1), false));
+        assertThatrangeSlicesWeigherReturnsCorrectNumRows(ThriftQueryWeighers.getRangeSlices(new KeyRange(1), true));
+    }
+
+    private void assertThatrangeSlicesWeigherReturnsCorrectNumRows(
```
Comment (jeremyk-91) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighersTest.java]: nit: `Range` (capitalisation)

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighersTest.java [hunk 4]
```diff
@@ -77,26 +97,54 @@ public void multigetSliceWeigherReturnsCorrectNumRows() {
...
+                .estimate().numBytes();
+        long numBytesWithTwoRows = ThriftQueryWeighers.getRangeSlices(new KeyRange(2), zeroEstimate)
+                .estimate().numBytes();
+
+        if (zeroEstimate) {
+            assertThat(numBytesWithOneRow).isEqualTo(0L);
+        } else {
+            assertThat(numBytesWithOneRow).isGreaterThan(0L);
+        }
+        assertThat(numBytesWithTwoRows).isEqualTo(zeroEstimate ? 0L : numBytesWithOneRow * 2);
```
Comment (jeremyk-91) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighersTest.java]: nit: this block looks similar to the one at the end of `assertThatEstimatesAreCorrect` - we might want to extract a method?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/QosCassandraClient.java (hunk 1 above)
Comment (hsaraogi) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/QosCassandraClient.java]: backup lock goes through this code path and actually releases a not of consumed bytes. I think it shouldn't be rate-limited, but didnt see enough evidence(see it fail due to the limit exeeding) in the tests to exclude it.

```
get_range_slices on table _persisted_locks for keyRange KeyRange(start_key:, end_key:, count:100)
Consuming an estimated bytes of 10000
Now time is 394
Wait time is 419
Recording an adjusted bytes of -9850
```

Also, note that the transaction table is not a hidden table, so HiddenTables::isHidden is not sufficient.

Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/QosCassandraClient.java]: Ah, you're right. If we wanted to do this programmatically we should probably check against the `hiddenTables` from AtlasDbConstants actually... (which still doesn't seem to have the default metadata table)

In any case this is fine as it is :)

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighersTest.java (hunk 4 above)
Comment (hsaraogi) [reply atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighersTest.java]: Done.

**Pull request merged**